source/cpu_usage_sampler.c

```diff
@@ -0,0 +1,238 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/cpu_usage_sampler.h>
+
+#if defined(__linux__) || defined(__unix__)
+#    include <sys/sysinfo.h>
+#    include <sys/types.h>
+#endif
+
+/*********************************************************************************************************************
+ * Base implementation
+ ********************************************************************************************************************/
+
+struct aws_cpu_sampler_vtable {
+    int (*aws_get_cpu_sample_fn)(struct aws_cpu_sampler *sampler, double *output);
+    void (*aws_cpu_sampler_destroy)(struct aws_cpu_sampler *sampler);
+};
+
+struct aws_cpu_sampler {
+    const struct aws_cpu_sampler_vtable *vtable;
+    struct aws_allocator *allocator;
+    void *impl;
+};
+
+static void aws_cpu_sampler_destroy_default(struct aws_cpu_sampler *sampler) {
+    if (sampler == NULL) {
+        return;
+    }
+    aws_mem_release(sampler->allocator, sampler);
+}
+
+static struct aws_cpu_sampler_vtable aws_cpu_sampler_vtable_default = {
+    .aws_get_cpu_sample_fn = NULL,
+    .aws_cpu_sampler_destroy = aws_cpu_sampler_destroy_default,
+};
+
+/*********************************************************************************************************************
+ * Linux
+ ********************************************************************************************************************/
+
+#if defined(__linux__) || defined(__unix__)
+
+struct aws_cpu_sampler_linux {
+    struct aws_cpu_sampler base;
+
+    uint64_t cpu_last_total_user;
+    uint64_t cpu_last_total_user_low;
+    uint64_t cpu_last_total_system;
+    uint64_t cpu_last_total_idle;
+    // CPU usage is reported in deltas on Linux.
+    // For the first report we have to process and cache but not send.
+    bool cpu_only_gather;
+};
+
+static void s_get_cpu_usage_linux(
+    uint64_t *total_user,
+    uint64_t *total_user_low,
+    uint64_t *total_system,
+    uint64_t *total_idle) {
+
+    FILE *file;
+    int matched_results;
+    file = fopen("/proc/stat", "r");
+    matched_results = fscanf(
+        file,
+        "cpu %llu %llu %llu %llu",
```

Comment:
I think we want to use `SCNu64` here rather than `llu`.  

It's a macro so you need to reference it outside the string constant:

```
fscanf(file, "cpu %" SCNu64 " %" SCNu64 " %" SCNu64" %" SCNu64, ....); 
```

See how we use PRIu64 elsewhere in common.

https://en.cppreference.com/w/c/types/integer